x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/rescorer/QueryFeatureExtractor.java

```diff
@@ -0,0 +1,117 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.ml.inference.rescorer;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.DisiPriorityQueue;
+import org.apache.lucene.search.DisiWrapper;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class QueryFeatureExtractor implements FeatureExtractor {
+
+    private final List<String> featureNames;
+    private final List<Weight> weights;
+    private final List<Scorer> scorers;
+    private DisjunctionDISI rankerIterator;
+
+    public QueryFeatureExtractor(List<String> featureNames, List<Weight> weights) {
+        this.featureNames = featureNames;
+        this.weights = weights;
+        this.scorers = new ArrayList<>(weights.size());
+    }
+
+    @Override
+    public void setNextReader(LeafReaderContext segmentContext) throws IOException {
+        DisiPriorityQueue disiPriorityQueue = new DisiPriorityQueue(weights.size());
+        scorers.clear();
+        for (Weight weight : weights) {
+            if (weight == null) {
+                scorers.add(null);
+                continue;
+            }
+            Scorer scorer = weight.scorer(segmentContext);
+            if (scorer != null) {
+                disiPriorityQueue.add(new DisiWrapper(scorer));
+            }
+            scorers.add(scorer);
+        }
+        rankerIterator = new DisjunctionDISI(DocIdSetIterator.all(segmentContext.reader().maxDoc()), disiPriorityQueue);
+    }
```

Comment:
Is there a reason why we need an iterator over all documents of the segment? Why we can't build an iterator over only `disiPriorityQueue` , as done in for example [here](https://github.com/apache/lucene/blob/main/lucene/core/src/java/org/apache/lucene/search/DisjunctionDISIApproximation.java)?